components/movementsensor/dualgps/dualgps.go

```diff
@@ -0,0 +1,236 @@
+	}
+	deps = append(deps, c.Gps1)
+
+	if c.Gps2 == "" {
+		return nil, resource.NewConfigValidationFieldRequiredError(path, "second_gps")
+	}
+	deps = append(deps, c.Gps2)
+
+	return deps, nil
+}
+
+func init() {
+	resource.RegisterComponent(
+		movementsensor.API,
+		model,
+		resource.Registration[movementsensor.MovementSensor, *Config]{Constructor: newDualGPS})
+}
+
+type dualGPS struct {
+	resource.Named
+	logger logging.Logger
+	mu     sync.Mutex
+	offset float64
+
+	gps1 gpsWithLock
+	gps2 gpsWithLock
+}
+
+// newDualGPS makes a new fake movement sensor.
+func newDualGPS(ctx context.Context, deps resource.Dependencies, conf resource.Config, logger logging.Logger,
+) (movementsensor.MovementSensor, error) {
+	dg := dualGPS{
+		Named:  conf.ResourceName().AsNamed(),
+		logger: logger,
+	}
+
+	if err := dg.Reconfigure(ctx, deps, conf); err != nil {
+		return nil, err
+	}
+	return &dg, nil
+}
+
+func (dg *dualGPS) Reconfigure(ctx context.Context, deps resource.Dependencies, conf resource.Config) error {
+	newConf, err := resource.NativeConfig[*Config](conf)
+	if err != nil {
+		return err
+	}
+
+	dg.mu.Lock()
+	defer dg.mu.Unlock()
+
+	first, err := movementsensor.FromDependencies(deps, newConf.Gps1)
+	if err != nil {
+		return err
+	}
+	dg.gps1.gps = first
+
+	second, err := movementsensor.FromDependencies(deps, newConf.Gps2)
+	if err != nil {
+		return err
+	}
+	dg.gps2.gps = second
+
+	dg.offset = defaultOffsetDegrees
+	if newConf.Offset != 0 {
+		dg.offset = newConf.Offset
+	}
+
+	return nil
+}
+
+type gpsWithLock struct {
+	gps     movementsensor.MovementSensor
+	gpsLock sync.Mutex
+}
+
+func (lgps *gpsWithLock) getPosition(ctx context.Context, extra map[string]interface{}) (*geo.Point, error) {
+	lgps.gpsLock.Lock()
+	defer lgps.gpsLock.Unlock()
+	pos, _, err := lgps.gps.Position(ctx, extra)
+	return pos, err
+}
+
+// Position gets the position of a fake movementsensor.
+func (dg *dualGPS) Position(ctx context.Context, extra map[string]interface{}) (*geo.Point, float64, error) {
+	return geo.NewPoint(math.NaN(), math.NaN()), math.NaN(), movementsensor.ErrMethodUnimplementedPosition
+}
+
+// getHeading calculates bearing and absolute heading angles given 2 geoPoint coordinates
+// 0 degrees is North, 90 degrees is East, 180 degrees is South, 270 is West.
+func getHeading(first, second *geo.Point, yawOffset float64) (float64, float64, float64) {
+	// convert latitude and longitude readings from degrees to radians
+	// so we can use go's periodic math functions
+	firstLat := utils.DegToRad(first.Lat())
+	firstLong := utils.DegToRad(first.Lng())
+	secondLat := utils.DegToRad(second.Lat())
+	secondLong := utils.DegToRad(second.Lng())
+
+	// calculate the bearing between gps1 and gps2
+	deltaLong := secondLong - firstLong
+	y := math.Sin(deltaLong) * math.Cos(secondLat)
+	x := math.Cos(firstLat)*math.Sin(secondLat) - math.Sin(firstLat)*math.Cos(secondLat)*math.Cos(deltaLong)
+	bearing := utils.RadToDeg(math.Atan2(y, x))
```

Comment:
Is there a reference URL to where these calculations came from? I trust you that it's right, but don't know how to verify that.